### PR TITLE
 UITrait: make it possible to pass a custom layout

### LINKED ISSUE #129: Remove the default layout set from the trait
I suggest to remove the lines:
```PHP
if (empty($this->layout) || $this->layout === 'default') {
    $this->layout = 'BootstrapUI.default';
}
```
from `initializeUI()` in the `UIViewTrait`, because doing so you kind of force the user to create another template name and set it into its own `AppView`.
Or change the method to support some args and then enable, like:
```PHP
public function initializeUI(array $args)
{
    if ($args['useBsLayout'] && (empty($this->layout) || $this->layout === 'default')) {
        $this->layout = 'BootstrapUI.default';
    }
}
```
and inside your AppView you pass what you need: `$this->initializeUI(['useBsLayout' => true]);`.
=)

### PR DESCRIPTION
**Pull request by ayman-alkom**
Fix #129

### CODE REVIEW

File: src/View/UIView.php [hunk 1]
```diff
@@ -17,12 +17,14 @@ class UIView extends View
     /**
      * Initialization hook method.
      *
+     * @param array $options
      * @return void
      */
-    public function initialize()
+    public function initialize(array $options = [])
```
Comment (ADmad) [inline src/View/UIView.php]: `View::initialize()` does not take any params, so this proposed change will generate errors.

Comment (ayman-alkom) [inline src/View/UIView.php]: yes that's right

Comment (ayman-alkom) [inline src/View/UIView.php]: then we can just use the trait for the custom layout, and the class will always use the plugin's one

File: src/View/UIView.php [hunk 2]
```diff
@@ -17,12 +17,14 @@ class UIView extends View
     /**
      * Initialization hook method.
      *
+     * @param array $options
```
Comment (ADmad) [inline src/View/UIView.php]: This needs to be removed

File: src/View/UIView.php [hunk 3]
```diff
@@ -17,12 +17,14 @@ class UIView extends View
...
      * @return void
      */
     public function initialize()
     {
         parent::initialize();
 
+        //render the initializeUI method from the UIViewTrait
```
Comment (ADmad) [inline src/View/UIView.php]: Please don't add inline comments

File: src/View/UIViewTrait.php [hunk 4]
```diff
@@ -3,22 +3,35 @@
...
+     *
+     * @param array $options
      * @return void
      */
-    public function initializeUI()
+    public function initializeUI(array $options = [])
     {
-        if (empty($this->layout) || $this->layout === 'default') {
-            $this->layout = 'BootstrapUI.default';
+        if ($options['layout'] !== false) {
```
Comment (ADmad) [inline src/View/UIViewTrait.php]: This will generate error if `layout` key is not set.

File: src/View/UIViewTrait.php [hunk 5]
```diff
@@ -3,22 +3,35 @@
...
+            if (is_string($options['layout']))
+                $this->layout = $options['layout'];
+            else
+                $this->layout = 'BootstrapUI.default';
         }
-
+        //Load BootstrapUI's helpers
```
Comment (ADmad) [inline src/View/UIViewTrait.php]: Remove inline comment

**Comment by ADmad** [comment]
Fix the various coding standard errors as reported [here](https://travis-ci.org/FriendsOfCake/bootstrap-ui/jobs/110970894#L312)

**Comment by ayman-alkom** [comment]
Hi @ADmad , what I have to do for codecov issue?

**Comment by ADmad** [comment]
Add tests for UIView and UIViewTrait. Currently then don't have any coverage.

https://codecov.io/github/FriendsOfCake/bootstrap-ui/src/View/UIView.php?ref=5b9d15427ca51bd5277907e4cce92797c9350a06
https://codecov.io/github/FriendsOfCake/bootstrap-ui/src/View/UIViewTrait.php?ref=5b9d15427ca51bd5277907e4cce92797c9350a06

File: README.md [hunk 6]
```diff
@@ -60,6 +60,7 @@ class AppView extends UIView
     public function initialize()
     {
         //Don't forget to call the parent::initialize()
+        //parent::initialize(['layout' => 'your_custom_layout']);
```
Comment (ADmad) [inline README.md]: This is not needed.

**Comment by ayman-alkom** [comment]
ok :(

**Comment by ayman-alkom** [comment]
Hi @ADmad it has been finished finally

File: src/View/UIViewTrait.php [hunk 7]
```diff
@@ -3,20 +3,37 @@
...
-        if (empty($this->layout) || $this->layout === 'default') {
+        if (!array_key_exists('layout', $options)) {
+            $this->layout = 'BootstrapUI.default';
+        } elseif ($options['layout'] === true) {
             $this->layout = 'BootstrapUI.default';
+        } elseif (is_string($options['layout'])) {
+            $this->layout = $options['layout'];
+        }
+
+        $layout = (is_string(@$options['layout'])) ? $options['layout'] : 'BootstrapUI.default';
```
Comment (ADmad) [inline src/View/UIViewTrait.php]: I think line 34-37 are not needed and should be removed. All cases are already covered above.

File: tests/TestCase/View/UIViewTraitTest.php [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    public function testInitializeUI()
+    {
+        $this->View->initializeUI();
+        $this->assertEquals('BootstrapUI.default', $this->View->layout);
+
+        $this->View->initializeUI([
+            'layout' => true
+        ]);
+        $this->assertEquals('BootstrapUI.default', $this->View->layout);
+//
```
Comment (ADmad) [inline tests/TestCase/View/UIViewTraitTest.php]: Needs to be removed.

**Comment by ayman-alkom** [comment]
It took a lot to be performed, because this is the first test I made, it is more simple than I thought

**Comment by ayman-alkom** [comment]
Hi @ADmad I removed them

File: src/View/UIViewTrait.php [hunk 9]
```diff
@@ -3,20 +3,29 @@
...
+     *      - If false do nothing ( will keep your layout )
+     *
+     * @return  void
      */
-    public function initializeUI()
+    public function initializeUI(array $options = [])
     {
-        if (empty($this->layout) || $this->layout === 'default') {
-            $this->layout = 'BootstrapUI.default';
+        $layout = (is_string(@$options['layout'])) ? $options['layout'] : 'BootstrapUI.default';
```
Comment (ADmad) [inline src/View/UIViewTrait.php]: You did the opposite of what I told you :(

Comment (ADmad) [inline src/View/UIViewTrait.php]: Use error silencing operator `@` is bad coding.

Comment (ADmad) [inline src/View/UIViewTrait.php]: Replace all your `$this->layout` assignment code with this:

```
        if (!isset($options['layout']) || $options['layout'] === true) {
            $this->layout = 'BootstrapUI.default';
        } elseif (!empty($options['layout'])) {
            $this->layout = $options['layout'];
        }
```

File: tests/TestCase/View/UIViewTraitTest.php [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        $this->View->initializeUI([
+            'layout' => 'myLayout'
+        ]);
+        $this->assertEquals('myLayout', $this->View->layout);
+
+        $this->View->layout = 'other_layout';
+        $this->View->initializeUI([
+            'layout' => false
+        ]);
+        $this->assertEquals('other_layout', $this->View->layout);
```
Comment (ADmad) [inline tests/TestCase/View/UIViewTraitTest.php]: Also add a test case for `'layout' => ''`.

File: src/View/UIViewTrait.php (hunk 9 above)
Comment (ayman-alkom) [inline src/View/UIViewTrait.php]: Yes I know ;)
You are right about the bad of using `@`, but I thought that more clean, I will retrieve them.

File: tests/TestCase/View/UIViewTraitTest.php (hunk 10 above)
Comment (ayman-alkom) [inline tests/TestCase/View/UIViewTraitTest.php]: ok

**Comment by ayman-alkom** [comment]
@ADmad  take a look

File: src/View/UIViewTrait.php [hunk 11]
```diff
@@ -3,20 +3,32 @@
...
+     *      - If a layout name passed it will be used
+     *      - If false do nothing ( will keep your layout )
+     *
+     * @return  void
      */
-    public function initializeUI()
+    public function initializeUI(array $options = [])
     {
-        if (empty($this->layout) || $this->layout === 'default') {
+        if (!array_key_exists('layout', $options)) {
```
Comment (ADmad) [inline src/View/UIViewTrait.php]: You are really testing my patience. I gave the exact code to use and yet you did something different.

File: tests/TestCase/View/UIViewTraitTest.php [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+        $this->View->layout = 'default';
+        $this->View->initializeUI([
+            'layout' => ''
+        ]);
+        $this->assertEquals('BootstrapUI.default', $this->View->layout);
```
Comment (ADmad) [inline tests/TestCase/View/UIViewTraitTest.php]: If layout option is empty string then view layout should also be empty string as its a valid value.

File: src/View/UIViewTrait.php (hunk 11 above)
Comment (ayman-alkom) [inline src/View/UIViewTrait.php]: sorry for that @ADmad , you may do it by yourself as you want

**Pull request closed without merging**